src/server/client.ts

```diff
@@ -0,0 +1,505 @@
+                this.fileMapping[(<ts.server.protocol.IdFile>fileId).id] = fileName;
+            }
+            else if (typeof fileId === "number") {
+                fileName = ts.lookUp(this.fileMapping, fileId.toString());
+                Debug.assert(!!fileName, "Did not find filename in previous fileID mappings.");
+            }
+            else {
+                Debug.fail("Got unexpedted fileId type.");
+            }
+            return fileName;
+        }
+
+        private processRequest<T extends ts.server.protocol.Request>(command: string, arguments?: any): T {
+            var request: ts.server.protocol.Request = {
+                seq: this.sequence++,
+                type: "request",
+                command: command,
+                arguments: arguments
+            };
+
+            this.writeMessage(JSON.stringify(request));
+
+            return <T>request;
+        }
+
+        private processResponse<T extends ts.server.protocol.Response>(request: ts.server.protocol.Request): T {
+            var lastMessage = this.messages.shift();
+            Debug.assert(!!lastMessage, "Did not recieve any responses.");
+
+            // Read the content length
+            var contentLengthPrefix = "Content-Length: ";
+            var lines = lastMessage.split("\r\n");
+            Debug.assert(lines.length >= 2, "Malformed response: Expected 3 lines in the response.");
+
+            var contentLengthText = lines[0];
+            Debug.assert(contentLengthText.indexOf(contentLengthPrefix) === 0, "Malformed response: Response text did not contain content-length header.");
+            var contentLength = parseInt(contentLengthText.substring(contentLengthPrefix.length));
+
+            // Read the body
+            var responseBody = lines[2];
+
+            // Verify content length
+            Debug.assert(responseBody.length + 1 === contentLength, "Malformed response: Content length did not match the response's body length.");
+
+            try {
+                var response: T = JSON.parse(responseBody);
+            }
+            catch (e) {
+                throw new Error("Malformed response: Failed to parse server response: " + lastMessage + ". \r\n  Error detailes: " + e.message);
+            }
+
+            // verify the sequence numbers
+            Debug.assert(response.request_seq === request.seq, "Malformed response: response sequance number did not match request sequence number.");
+            
+            // unmarshal errors
+            if (!response.success) {
+                throw new Error("Error " + response.message);
+            }
+
+            Debug.assert(!!response.body, "Malformed response: Unexpected empty response body.");
+
+            return response;
+        }
+
+        openFile(fileName: string): void {
+            var args: ts.server.protocol.FileRequestArgs = { file: fileName };
+            this.processRequest(CommandNames.Open, args);

```

Comment:
Command names should bind much more tightly to our actual operations.  'open' and 'close' are very vague.  They should be OpenFile and CloseFile.